server/app/services/program/ActiveAndDraftPrograms.java

```diff
@@ -70,30 +83,109 @@ private ImmutableMap<String, ProgramDefinition> mapNameToProgram(
+        mapNameToProgramWithFilter(repository, service, active, Optional.of(DisplayMode.DISABLED));
+
+    // All active programs (including disabled).
+    ImmutableMap<String, ProgramDefinition> activeNameToProgramAll =
         mapNameToProgram(repository, service, active);
 
+    // Draft, non-disabled programs.
     ImmutableMap<String, ProgramDefinition> draftNameToProgram =
+        mapNameToProgramWithFilter(repository, service, draft, Optional.of(DisplayMode.DISABLED));
+
+    // All draft programs (including disabled).
+    ImmutableMap<String, ProgramDefinition> draftNameToProgramAll =
         mapNameToProgram(repository, service, draft);
 
-    this.activePrograms = activeNameToProgram.values().asList();
-    this.draftPrograms = draftNameToProgram.values().asList();
-    this.versionedByName =
-        Sets.union(activeNameToProgram.keySet(), draftNameToProgram.keySet()).stream()
-            .collect(
-                ImmutableMap.toImmutableMap(
-                    Function.identity(),
-                    programName -> {
-                      return Pair.create(
-                          Optional.ofNullable(activeNameToProgram.get(programName)),
-                          Optional.ofNullable(draftNameToProgram.get(programName)));
-                    }));
+    if (types.size() == 2
+        && types.contains(ActiveAndDraftProgramsType.DISABLED)
+        && types.contains(ActiveAndDraftProgramsType.IN_USE)) {
+      this.activePrograms = activeNameToProgramAll.values().asList();
+      this.draftPrograms = draftNameToProgramAll.values().asList();
+      this.versionedByName =
+          createVersionedByNameMap(activeNameToProgramAll, draftNameToProgramAll);
+    } else if (types.size() == 1) {
+      ActiveAndDraftProgramsType type = types.get(0);
+      if (type.equals(ActiveAndDraftProgramsType.DISABLED)) {
+        this.activePrograms = activeNameToProgram.values().asList();
+        this.draftPrograms = draftNameToProgram.values().asList();
+        // Disabled active programs.
+        ImmutableMap<String, ProgramDefinition> disabledActiveNameToProgram =
+            filterMapNameToProgram(activeNameToProgramAll, activeNameToProgram);
+        // Disabled draft programs.
+        ImmutableMap<String, ProgramDefinition> disabledDraftNameToProgram =
+            filterMapNameToProgram(draftNameToProgramAll, draftNameToProgram);
+        this.versionedByName =
+            createVersionedByNameMap(disabledActiveNameToProgram, disabledDraftNameToProgram);
+      } else if (type.equals(ActiveAndDraftProgramsType.IN_USE)) {
+        this.activePrograms = activeNameToProgram.values().asList();
+        this.draftPrograms = draftNameToProgram.values().asList();
+        this.versionedByName = createVersionedByNameMap(activeNameToProgram, draftNameToProgram);
+      } else {
+        throw new IllegalArgumentException("Unsupported ActiveAndDraftProgramsType: " + type);
+      }
+    } else {
```

Comment:
**suggestion:** Before the other if condition could instead consider saying if types.size > 3, so we don't have to check the size at each other step.

Reply:
Would you mind clarifying which `if` condition you're referring to?

Reply:
Not sure which `if` is referred here but made some changes given my best guess. Let me know if that's not what you meant.

Reply:
Sorry for the confusion. I meant that instead of having a nested if for types.size < 3 we could remove that if condition all together. If we wanted to make sure only ones with < 3 got to this point, we could create a separate if condition above it for > 3. But with `EnumSet` (mentioned above) we shouldn't need to check the size.